src/infql.jl

```diff
@@ -378,3 +378,216 @@ function LazyBandedMatrices._SymTridiagonal(::Tuple{TriangularLayout{'L', 'N', P
+        Lerr = norm(Ll-Ls,2)
+        if N == maxN
+            error("Reached max. iterations in finite section QL without convergence to desired tolerance.")
+        end
+        Ls = Ll
+        N = 2*N
+    end
+    A.data = ql(A.M[1:(N÷2),1:(N÷2)]).τ[1:j]
+end
+
+function getindex(K::AdaptiveQLFactors, k::Integer, j::Integer)
+    resizedata!(K, max(k,j))
+    K.data[k, j]
+end
+function getindex(K::AdaptiveQLFactors, k::Integer, jr::UnitRange{Int})
+    resizedata!(K, max(k,maximum(jr)))
+    K.data[k, jr]
+end
+function getindex(K::AdaptiveQLFactors, kr::UnitRange{Int}, j::Integer)
+    resizedata!(K, max(j,maximum(kr)))
+    K.data[kr, j]
+end
+function getindex(K::AdaptiveQLFactors, kr::UnitRange{Int}, jr::UnitRange{Int})
+    resizedata!(K, max(maximum(jr),maximum(kr)))
+    K.data[kr, jr]
+end
+function getindex(K::AdaptiveQLTau, jr::UnitRange{Int})
+    resizedata!(K, maximum(jr))
+    K.data[jr]
+end
+function getindex(K::AdaptiveQLTau, k::Integer)
+    resizedata!(K, k)
+    K.data[k]
+end
+
+function getproperty(F::QL{T, AdaptiveQLFactors{T}, AdaptiveQLTau{T}}, d::Symbol) where T
+    if d == :L
+        return LowerTriangular(F.factors)
+    elseif d == :Q
+        return getQ(F)
+    else
+        getfield(F, d)
+    end
+end
+
+@inline function getQ(F::QL{T, AdaptiveQLFactors{T}, AdaptiveQLTau{T}}) where T
+    Q = AdaptiveQLQ{T}([one(T)],F.factors,F.τ,(0,0))
+    # initialize as 50x50, then adaptively expand
+    resizedata!(Q,50)
+    return Q
+end
+
+mutable struct AdaptiveQLQ{T} <: AbstractCachedMatrix{T}
+    data
+    factors::AdaptiveQLFactors{T}
+    τ::AdaptiveQLTau{T}
+    datasize::Tuple{Int, Int}
+    AdaptiveQLQ{T}(D, f, tau, N::Tuple{Int, Int}) where T = new{T}(D, f, tau, N)
+end
+
+size(::AdaptiveQLQ) = (ℵ₀, ℵ₀)
+
+function resizedata!(K::AdaptiveQLQ, nm::Integer)
+    νμ = K.datasize[1]
+    if nm > νμ
+        inds = νμ:nm
+        cache_filldata!(K, inds)
+        K.datasize = size(K.data).÷2
+    end
+    K
+end
+
+# Q = \\prod_{i=1}^{\\min(m,n)} (I - \\tau_i v_i v_i^T)
+function cache_filldata!(F::AdaptiveQLQ{T}, inds::UnitRange{Int}) where T
+    nm = 2*maximum(inds)
+    τ = F.τ[1:nm]
+    v = I + triu(F.factors[1:nm,1:nm],1)
+    prod = Diagonal(ones(T,nm))
+    @inbounds for j = 1:nm
+        prod = prod * (I - F.τ[j]*v[:,j]*v[:,j]')
+    end
+    F.data = prod'
+end
+
+function getindex(K::AdaptiveQLQ, k::Integer, j::Integer)
+    resizedata!(K, max(k,j))
+    K.data[k, j]
+end
+function getindex(K::AdaptiveQLQ, k::Integer, jr::UnitRange{Int})
+    resizedata!(K, max(k,maximum(jr)))
+    K.data[k, jr]
+end
+function getindex(K::AdaptiveQLQ, kr::UnitRange{Int}, j::Integer)
+    resizedata!(K, max(j,maximum(kr)))
+    K.data[kr, j]
+end
+function getindex(K::AdaptiveQLQ, kr::UnitRange{Int}, jr::UnitRange{Int})
+    resizedata!(K, max(maximum(jr),maximum(kr)))
+    K.data[kr, jr]
+end
+
+*(L::LowerTriangular{T, AdaptiveQLFactors{T}}, b::LayoutVector) where T = ApplyArray(*, L, b)
```

Comment:
It would be better to adaptively build `L*b` using caching and forward-substitution. Though perhaps it can stay for now since that would need a special type. But put a `#TODO` to remind us to change this.

Reply:
will do